docs/content/docs/data-structures/tree.md

```diff
@@ -485,61 +498,98 @@ For the meaning of "simultaneously", see [Types of distributed data structures](
 
 ### Events
 
-The `SharedTree` object supports two events: `beforeChange` and `afterChange`. Your code can create handlers for these events using the utility class `Tree`. See [Tree utility APIs](#tree-utility-apis).
+`SharedTree` supports two node level events: `nodeChanged` and `treeChanged`. Your code can create handlers for these events using the utility class `Tree`. See [Tree utility APIs](#tree-utility-apis).
+
+Additionally, the `TreeView` object includes 3 events that operate over the whole tree. These are `rootChanged`, `afterBatch`, and `commitApplied`.
+
+`rootChanged` fires when the root field (the field that contains the root node) changes. That is, if a new root node is assigned or the schema changes. This will not fire when the node itself changes.
+
+`afterBatch` fires when a batch of changes has been processed by the Fluid Framework and the `TreeView` object is updated.
+
+`commitApplied` fires whenever a local change is applied outside of a transaction or when a local transaction is committed. This is used to get `Revertible` objects to put on the undo stack. See [Undo/Redo support](#undoredo-support) and [Transactions](#transactions).
+
+### Undo/Redo support
+
+`SharedTree` makes creating an undo and redo stack very simple. By listening for the `commitApplied` event on the `TreeView` object, you can get a `Revertible` object.
+`Revertible` objects come in three flavors:
+
+-   Default: a normal change made in the local client that would go on the undo stack
+-   Undo: a change that is the result of reverting a Default or Redo `Revertible` object that would go on the redo stack
+-   Redo: a change that is the result of reverting an Undo `Revertible` object that would go on the undo stack
+
+To undo a change, call the `revert` method on the `Revertible` object. This will return the `TreeView` object to the state last observed by the local client. If changes were made by other clients that are incompatible with the state of the tree after undo,
+these changes will also be reverted. Calling `revert` on the Redo `Revertible` object will put everything back to the way it was if the change was unintentially destructive.
```

Comment:
There are a few issues here:
1. "return the `TreeView` object to the state last observed by the local client" sounds like (or at least can plausibly be interpreted as) rewinding to whichever state the view was in before the last applied change. Instead, it seems better to say that this reverts the properties were edited by the reverted commit to the state that there were in on the local client before the reverted commit.
2. "If changes were made by other clients..." is confusing because we don't define what "incompatible" means in this context. An example would be good.
3. "these changes will also be reverted" is not quite right by our definition of what revert means. Instead, I would just say that they will be overwritten.
4. "will put everything back to the way it was": the "everything" is wrong because other changes could have been applied (concurrently or not) somewhere else in the tree and those would not be rolled back. The qualification ""
5. "if the change was unintentially destructive": this is tacked on to the description of what calling revert does, so it sounds as though your description of what it does is predicated on whether or not the change was <s>unintentially</s> unintentionally destructive. I'm guessing this part was meant to illustrate why you'd want to redo, but I don't think it's needed here (or it should at least be separate).